### ISSUE TITLE
Source code fixer doesn't remove sequential ranges

### ISSUE THREAD

**Issue by pantosha** [eslint/eslint#10601]
**Tell us about your environment**

* **ESLint Version:** 5.1.0
* **Node Version:** 10.6.0
* **npm Version:** 6.1.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default
**Please show your full configuration:**
I cloned eslint repo.

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

I'm developing fixer for `no-unused-var` rule. Fixer should remove unused method parameters. Fixer removes one parameter (sometimes with comma) per error. Error is reported for each unused argument.
`SourceCodeFixer` should successfully remove sequential ranges, but `SourceCodeFixer` recognizes them as overlaps.

I wrote test for `SourceCodeFixer`:
```js
            const REMOVE_A = {
                message: "removea",
                fix: {
                    range: [11, 12],
                    text: ""
                }
            };

            const REMOVE_B = {
                message: "removeb",
                fix: {
                    range: [12, 15],
                    text: ""
                }
            };

            const REMOVE_C = {
                message: "removec",
                fix: {
                    range: [15, 18],
                    text: ""
                }
            };

            it("should remove text consistently", () => {
                const result = SourceCodeFixer.applyFixes('function f(a, b, c) {}', [REMOVE_A, REMOVE_B, REMOVE_C]);

                assert.strictEqual(result.output, "function f() {}");
                assert.isTrue(result.fixed);
            });
```

**What did you expect to happen?**
Test should be successful. Output should be `"function f() {}"`.

**What actually happened? Please include the actual, raw output from ESLint.**
Test is failed. Output is `"function f(, b) {}"`.

**Comment by not-an-aardvark** [comment]
Hi, thanks for reporting an issue. It's worth noting that any given fix is not guaranteed to be applied by ESLint, so I think ESLint is behaving as documented here. When writing an autofixer for a rule, it's important to make sure that the resulting code is still valid even if only a subset of the given fixes are applied.

It's possible that we can make a change to apply more fixes than we did previously, but the example fix you provided here would still be incorrect because it would be assuming that all fixes get applied.

**Comment by pantosha** [comment]
@not-an-aardvark , @mysticatea, thank you for explorations.

I understand and agree that overlapped ranges can't be applied.
But I think that it'll be nice if consecutive ranges can be handled by ESLint.

I prepared more detailed explanation. Condition that each error should be standalone is applied.

```
function f(a, b, c) {};
|          |  |  |
0         11 14 17 
```

There three errors are reported, one error for each method parameter.
Errors have fixers:

```
const REMOVE_A = {
    message: "removea",
    fix: {
        range: [11, 14],
        text: ""
    }
};

const REMOVE_B = {
    message: "removeb",
    fix: {
        range: [14, 17],
        text: ""
    }
};

const REMOVE_C = {
    message: "removec",
    fix: {
        range: [15, 18],
        text: ""
    }
};
```

**Expect:**
I expect that two fixes will be applied - `removea`, `removeb`.
Fix for `removec` won't be applied due to overlap with `removeb`.

Result will be 
```
function f(c) {};
```

**Actual:**
`removea` and `removec` will be applied. SourceCodeFixer will skip `removeb` because `removeb` range start is the same as the end of `removea` range (but the end of the range isn't modified!).

**Comment by not-an-aardvark** [comment]
I'm not sure this is a big problem because ESLint runs rules multiple times when autofixing, so the variable `b` will simply be removed on the next pass when the rule gets run again.

It's true that it would be safe to apply all the fixes in this case, but I'm not sure it's safe to apply all the fixes in all other cases. I would rather be conservative here.

**Comment by pantosha** [comment]
@not-an-aardvark, I agree that it's not a big problem.

There is another inconvenient place. Consecutive fixes are merged good when error's `fix` method returns array or implemented as generator, but `SourceCodeFixer` doesn't apply consecutive fixes 😢

**Comment by mysticatea** [comment]
@pantosha There is a difference; The former handles the fix items which came from the same lint errors. The latter handles the fix items which came from the different lint errors. The latter is not safer than the former.

**Comment by pantosha** [comment]
@mysticatea, why one symbol between two different fixers is more safer?

**Comment by mysticatea** [comment]
@pantosha Because of testability mainly. About the `fix` function, it's the matter in a lint error in a rule. On the other hand, `SourceCodeFixer` handles fix items which came from many different rules such as core rules and plugin rules. These combinations are infinite. We have sometimes experienced the broken which is caused by the confliction between the autofix of two rules. In my memory, we should make the current behavior about consecutive ranges in order to fix problems such a confliction.

**Comment by nzakas** [comment]
@mysticatea should this issue be flagged as a bug?

**Comment by mysticatea** [comment]
Probably, no. In my understanding, it has been introduced for #4321.

**Comment by nzakas** [comment]
Does that mean we can close this issue?

**Comment by mysticatea** [comment]
Yes.
